### ISSUE TITLE
Missing ssh package makes git unusable

### ISSUE THREAD

**Issue by henkf** [jenkinsci/docker-agent#322]
### Jenkins and plugins versions report

No further details necessary. Using the latest image (3071.v7e9b_0dc08466-4) from docker hub.

### What Operating System are you using (both controller, and any agents involved in the problem)?

Dockerised environment.

### Reproduction steps

Use the current latest image (3071.v7e9b_0dc08466-4) of jenkins/agent in a Jenkins Cloud Setup.

### Expected Results

The image should be able to make use of git via ssh.

### Actual Results

Git cannot connect. Relevant error message: ssh not found.

### Anything else?

The ssh command was present up to the second to latest build of this image.

**Comment by nagyo** [comment]
Hi,

We are experiencing the same issue. The latest jdk11 image [3071.v7e9b_0dc08466-4](https://hub.docker.com/layers/jenkins/agent/3071.v7e9b_0dc08466-4-jdk11/images/sha256-147a6af86e0b77a7f533a9bc5f89c92bb936cc86575dc89dcb2c93a93c996995?context=explore) throws the following during a git checkout:
```
09:47:36  stderr: /home/jenkins/workspace/other/test@tmp/jenkins-gitclient-ssh8648290266809590591.sh-copy: 6: ssh: not found
09:47:36  fatal: Could not read from remote repository.
09:47:36  
09:47:36  Please make sure you have the correct access rights
09:47:36  and the repository exists.
09:47:36  
09:47:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2736)
09:47:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2111)
09:47:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:87)
09:47:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:623)
09:47:36  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:852)
09:47:36  	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:158)
09:47:36  	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:151)
09:47:36  	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
09:47:36  	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
09:47:36  	at hudson.remoting.Request$2.run(Request.java:376)
09:47:36  	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
09:47:36  	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
09:47:36  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
09:47:36  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
09:47:36  	at java.base/java.lang.Thread.run(Unknown Source)
09:47:36  	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to agent-003a4mijtsmzb
09:47:36  		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1784)
09:47:36  		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
09:47:36  		at hudson.remoting.Channel.call(Channel.java:1000)
09:47:36  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:143)
09:47:36  		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
09:47:36  		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
09:47:36  		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
09:47:36  		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
09:47:36  		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:129)
09:47:36  		at com.sun.proxy.$Proxy218.execute(Unknown Source)
09:47:36  		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1229)
09:47:36  		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1312)
09:47:36  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:129)
09:47:36  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:97)
09:47:36  		at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:84)
09:47:36  		at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
09:47:36  		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
09:47:36  		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
09:47:36  		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
09:47:36  		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
09:47:36  		at java.base/java.lang.Thread.run(Thread.java:829)
09:47:36  [Pipeline] }
09:47:36  [Pipeline] // stage
09:47:36  [Pipeline] }
09:47:36  [Pipeline] // node
09:47:36  [Pipeline] End of Pipeline
09:47:36  ERROR: Error cloning remote repo 'origin'
09:47:36  Finished: FAILURE
```

Reverting back to [3071.v7e9b_0dc08466-3-jdk11](https://hub.docker.com/layers/jenkins/agent/3071.v7e9b_0dc08466-3-jdk11/images/sha256-a854e8294fe2ccfc2460c52dc2a05f65d613a240b514bc27155835669e27381a?context=explore) resolves the issue.

Regards,

Gábor

**Comment by dduportal** [comment]
Hi @henkf Thanks for raising the issue and sorri for the inconvenience.

Many thanks @nagyo for confirming AND providing a solution!

This problem was caused by https://github.com/jenkinsci/docker-agent/pull/320/files#diff-3f163a09ffba37e91b3af585fae100b7332047433e42b848836c4a492d10b6d7R53 (by me, sorry).

Incoming PR today (followed by a release) to add the needed packages.

If I may ask @henkf @nagyo is there a reason why you are using this image compared to the `jenkins/inbound-agent` image?

**Comment by henkf** [comment]
Thanks for the quick reaction, @dduportal. I'm looking forward to switch back to jenkins/agent:latest after your fix.
I would describe my setup similar to what @nagyo already explained. One node, throwaway build envs defined in container images, private git.
If you find the time to explain what makes the inbound agent a better fit I will be happy to reconsider my setup. Until then, using jenkins/agent seems to be the most comfortable way and I think it should remain capable of connecting to git repositories via ssh.

**Comment by dduportal** [comment]
(for information: PR opened by @gounthar : https://github.com/jenkinsci/docker-agent/pull/326)

**Comment by dduportal** [comment]
First of all, many thanks for your feedbacks @henkf @nagyo . It is very valuable to understand how are these images used!

Both your setup make sense (as soon as you never use "docker in docker" in these cases) .

So what you are using are inbound agent, without knowing it in fact: the Docker Cloud spawns agent process in the containers freshly created, and these agent processes connect to your controller (with a seuciryt token usually): hence the "inbound" naming.
It is opposed to the "SSH agent" (https://github.com/jenkinsci/docker-ssh-agent for instance) where the Jenkins controller connects to the remote system with SSH and takes care of getting the "agent.jar" file and spaws the agent process itself.

----

But why is there a difference between `jenkins/agent` and `jenkins/inbound-agent` ? I don't know why but:

- Most of the content in `jenkins/agent` (Java, bash, git, now SSH thanks to this issue, the default user `jenkins` with its UID `1000`, etc.) seems common to both "inbound agent" and "ssh agent" to me

- The "inbound-agent" image only adds a custom shell script as entrypoint: https://github.com/jenkinsci/docker-inbound-agent/blob/2ab4f9195c6ff3c98be2f283e002f8aca27ac78e/11/debian/Dockerfile#L31 (for debian JDK11 for instance) so it does not add much to `jenkins/agent` except the valuable script that can be tuned (using a different JDK, passing options to the agent process through env. vars etc.)

=> I understand that the Docker Cloud plugin does not seem to need this script as entrypoint: it calls the `java -jar /usr/share/jenkins/agent.jar` <parameters>` command to spawn the agent directly, which is why it works in your case.

Does my explanation make sense @henkf @nagyo (and also @timja @MarkEWaite @oleg-nenashev ?)

----

Why I'm asking these questions? Because I would liek to propose a "merge" of these 3 different images source codes into a single repository.
The reason is to ensure that we bump dependencies (java version, linux versions, remoting versions, etc.) only once for alland do the release once for all. Because most of the fixes such as this issue must be duplicated on each of the images (and I do not speak of the spread of testing cases...).

=> that would mean only providing 2 images: "inbound agent", and "ssh agent" in the future to simplify the life of consumers (and the documentation/example/etc). That would require changing the name one time one your own system to benefit from new upgrade (if it is done: it's only an idea) but would help the community. WDYT?

**Comment by dduportal** [comment]
Closing the issue as the latest release seems to fix the initial problem:

## Before:

```
docker run --rm -ti --entrypoint='' jenkins/agent:3071.v7e9b_0dc08466-4 git clone git@github.com:jenkinsci/docker-agent.git
Cloning into 'docker-agent'...
error: cannot run ssh: No such file or directory
fatal: unable to fork
```

## After

```
docker run --rm -ti --entrypoint='' jenkins/agent:3071.v7e9b_0dc08466-5 git clone git@github.com:jenkinsci/docker-agent.git
Status: Downloaded newer image for jenkins/agent:3071.v7e9b_0dc08466-5
Cloning into 'docker-agent'...
The authenticity of host 'github.com (140.82.121.4)' can't be established.
# ...
```


Please feel free to reopen if the issue is not fixed with the latest release!

(PS I'll open an issue about the "merge" images with my ramblings inside)

**Comment by nagyo** [comment]
Hey @dduportal,

Thanks again for the fix 👍🏻

I'm happy to test or experiment with the merged version of the inbound agent image, whenever it will be available. Quite familiar with the pain of maintaining the same thing in different locations, so feel free to tag me on the issue if any testing will be needed.

Regards,

Gábor

**Comment by dduportal** [comment]
> 



> Hey @dduportal,
> 
> Thanks again for the fix 👍🏻
> 
> I'm happy to test or experiment with the merged version of the inbound agent image, whenever it will be available. Quite familiar with the pain of maintaining the same thing in different locations, so feel free to tag me on the issue if any testing will be needed.
> 
> Regards,
> 
> Gábor

Many thanks @nagyo ! A good first step fro you would be to switch to the `jenkins/inbound-agent` image when ready.
(please note that the versions of this image are always a bit behind of the `jenkins/agent`: work is in progress with added automation, but I'm currently having issues with it so not ready right now).

**Comment by kubecto** [comment]
git clone http://xxx:xxx@devops.xxx.cn/gitlab/cs/x-cloud/script/deploy.git --branch develop --single-branch
Cloning into 'deploy'...
remote: HTTP Basic: Access denied
fatal: Authentication failed for

**Comment by gounthar** [comment]
What is the version of `git` you're using in other containers vs the one we're including?

**Comment by MarkEWaite** [comment]
> I can't pull the code this way, but I can pull the code in other containers, what is the cause of this
> 
> ```
> git clone http://xxx:xxx@devops.xxx.cn/gitlab/cs/x-cloud/script/deploy.git --branch develop --single-branch
> Cloning into 'deploy'...
> remote: HTTP Basic: Access denied
> fatal: Authentication failed for. 
> ```

@kubecto you're asking a question on an unrelated issue.  The clone that you are attempting to perform is using the `http` protocol.  This issue discusses and resolves an issue with the `ssh` protocol.  They are different protocols with different implementations.

It would be better if you opened a new issue rather than appending your question to this closed issue.

You're asking for help to diagnose an HTTP authentication failure with a custom `git clone` command.  Since the command you are using is not a command that is generated by the Jenkins git plugin, I assume you are performing that command yourself.  My first guess is that the username or password embedded in that URL are incorrect.  You could try placing that username and password into a Jenkins credential, then use the Jenkins git plugin to perform the operation.
